Complex enum variants with `Dir` causes error

Minimum repro:
```rust
#[pyclass]
#[derive(Dir)]
enum X {
    A {},
    B {
        x: bool,
    }
}
```
Error:
```
error[E0282]: type annotations needed
  --> mistralrs-pyo3/src/anymoe.rs:5:10
   |
 5 | #[derive(Dir)]
   |          ^^^ cannot infer type for array `[_; 0]`
   |
   = note: this error originates in the derive macro `Dir` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,3 +15,4 @@ repository = "https://github.com/EricLBuehler/pyo3-special-method-derive"
 
 [workspace.dependencies]
 pyo3 = { version = "0.22", features = ["multiple-pymethods"] }
+pyo3_special_method_derive_macro = { path = "pyo3_special_method_derive_macro" }

--- a/pyo3_special_method_derive/Cargo.toml
+++ b/pyo3_special_method_derive/Cargo.toml
@@ -7,5 +7,5 @@ license.workspace = true
 repository.workspace = true
 
 [dependencies]
-pyo3_special_method_derive_macro = "0.4.1"
+pyo3_special_method_derive_macro.workspace = true
 pyo3.workspace = true

--- a/pyo3_special_method_derive/tests/empty_complex_enum.rs
+++ b/pyo3_special_method_derive/tests/empty_complex_enum.rs
@@ -1,0 +1,8 @@
+use pyo3::pyclass;
+use pyo3_special_method_derive::Dir;
+
+#[pyclass]
+#[derive(Dir)]
+enum X {
+    A {},
+}

--- a/pyo3_special_method_derive_macro/src/lib.rs
+++ b/pyo3_special_method_derive_macro/src/lib.rs
@@ -135,19 +135,19 @@ pub fn dir_derive(input: TokenStream) -> TokenStream {
         }
         Data::Enum(e) => {
             let matchers = e.variants.iter()
-            .filter(|variant| {
-                !variant.attrs.iter().any(|attr| {
-                    let mut is_skip = false;
-                    if attr.path().is_ident(ATTR_NAMESPACE) { // only parse ATTR_NAMESPACE and not [serde] or [default]
-                        attr.parse_nested_meta(|meta| {
-                            is_skip = meta.path.is_ident("skip");
-                            Ok(())
-                        })
-                        .unwrap();
-                    }
-                    is_skip
+                .filter(|variant| {
+                    !variant.attrs.iter().any(|attr| {
+                        let mut is_skip = false;
+                        if attr.path().is_ident(ATTR_NAMESPACE) { // only parse ATTR_NAMESPACE and not [serde] or [default]
+                            attr.parse_nested_meta(|meta| {
+                                is_skip = meta.path.is_ident("skip");
+                                Ok(())
+                            })
+                            .unwrap();
+                        }
+                        is_skip
+                    })
                 })
-            })
                 .map(|variant| {
                     let ident = &variant.ident;
                     match &variant.fields {
@@ -161,19 +161,23 @@ pub fn dir_derive(input: TokenStream) -> TokenStream {
                         }
                         Fields::Named(fields) => {
                             let field_names = fields.named.iter().map(|f| f.ident.as_ref().unwrap().clone()).collect::<Vec<_>>();
-
-                            let mut assigner = proc_macro2::TokenStream::new();
-                            quote_into::quote_into!(assigner += [#{
-                                for name in &field_names {
-                                    quote_into::quote_into!(assigner += (names.push(stringify!(#name).to_string())),)
-                                }
-                            }];);
-
-                            quote! {
-                                Self::#ident { .. } => {
-                                    let mut names = Vec::new();
-                                    #assigner
-                                    names
+                            
+                            if field_names.is_empty() {
+                                quote! { Self::#ident { .. } => { vec![] } }
+                            } else {
+                                let mut assigner = proc_macro2::TokenStream::new();
+                                quote_into::quote_into!(assigner += [#{
+                                    for name in &field_names {
+                                        quote_into::quote_into!(assigner += (names.push(stringify!(#name).to_string())),)
+                                    }
+                                }];);
+    
+                                quote! {
+                                    Self::#ident { .. } => {
+                                        let mut names = Vec::new();
+                                        #assigner
+                                        names
+                                    }
                                 }
                             }
                         }
@@ -323,44 +327,6 @@ pub fn auto_display(input_stream: TokenStream) -> TokenStream {
 
         TokenStream::from(expanded)
     }
-}
-
-/// Implement `PyDisplay` and `Display` on a struct or enum.
-///
-/// This has the same requirements and behavior of [`Str`].
-///
-/// ## Example
-/// ```ignore
-/// use pyo3_special_method_derive::AutoDisplayDerive;
-/// #[derive(AutoDisplayDerive)]
-/// struct Person {
-///     pub name: String,
-///     address: String,
-///     #[auto_display(skip)]
-///     pub phone_number: String,
-///     #[auto_display] // -> force display of private field
-///     hash: u32,
-/// }
-/// ```
-#[proc_macro_derive(AutoDisplayDerive, attributes(pyo3_smd, pyo3_smd_str, auto_display))]
-pub fn auto_display_debug(input_stream: TokenStream) -> TokenStream {
-    let input = parse_macro_input!(input_stream as DeriveInput);
-    let name = &input.ident;
-
-    let display_debug_derive_body = impl_formatter(&input, DeriveType::ForAutoDisplay);
-
-    let expanded = quote! {
-        #display_debug_derive_body
-
-        impl std::fmt::Display for #name {
-            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-                use pyo3_special_method_derive::PyDisplay;
-                write!(f, "{}", self.fmt_display())
-            }
-        }
-    };
-
-    TokenStream::from(expanded)
 }
 
 /// Add a `__repr__` method to the struct or enum.

--- a/pyo3_special_method_derive_macro/src/lib.rs
+++ b/pyo3_special_method_derive_macro/src/lib.rs
@@ -161,7 +161,7 @@ pub fn dir_derive(input: TokenStream) -> TokenStream {
                         }
                         Fields::Named(fields) => {
                             let field_names = fields.named.iter().map(|f| f.ident.as_ref().unwrap().clone()).collect::<Vec<_>>();
-                            
+
                             if field_names.is_empty() {
                                 quote! { Self::#ident { .. } => { vec![] } }
                             } else {
@@ -171,7 +171,7 @@ pub fn dir_derive(input: TokenStream) -> TokenStream {
                                         quote_into::quote_into!(assigner += (names.push(stringify!(#name).to_string())),)
                                     }
                                 }];);
-    
+
                                 quote! {
                                     Self::#ident { .. } => {
                                         let mut names = Vec::new();

